src/script/redmine-upkeep.py

```diff
@@ -0,0 +1,886 @@
+        else:
+            issue_update.logger.info(f"Processing issue: {issue.url} '{issue.subject}'")
+
+        try:
+            applied_transformations = []
+            for transform_method in self.transform_methods:
+                try:
+                    # Each transformation method modifies the same issue_update object
+                    transform_name = transform_method.__name__.removeprefix("_transform_")
+                    issue_update.logger.debug(f"Calling transformation: {transform_name}")
+                    issue_update.set_transform(transform_name)
+                    if transform_method(issue_update):
+                        issue_update.logger.info(f"Transformation {transform_method.__name__} resulted in a change.")
+                        applied_transformations.append(transform_method.__name__)
+                except Exception as e:
+                    issue_update.logger.exception(f"Error applying transformation {transform_method.__name__}: {e}")
+                finally:
+                    issue_update.set_transform(None)
+
+            if issue_update.has_changes:
+                issue_update.logger.info("Changes detected. Sending update to Redmine...")
+                try:
+                    # We cannot put top-level changes in the PUT request against
+                    # the redmine API via redminelib. So we send it manually.
+                    payload = issue_update.get_update_payload()
+                    issue_update.logger.debug("PUT payload:\n%s", json.dumps(payload, indent=4))
+                    headers = {
+                        'Content-Type': 'application/json',
+                        'X-Redmine-API-Key': REDMINE_API_KEY,
+                    }
+                    endpoint = f"{REDMINE_ENDPOINT}/issues/{issue.id}.json"
+                    response = requests.put(endpoint, headers=headers, data=json.dumps(payload))
+                    response.raise_for_status()
+                    issue_update.logger.info("Successfully updated Redmine issue.")
+                    self.issues_modified += 1
+                    for t_name in applied_transformations:
+                        self.modifications_made.setdefault(t_name, 0)
+                        self.modifications_made[t_name] += 1
+                    return True
+                except requests.exceptions.HTTPError as err:
+                    issue_update.logger.error("API PUT failure during upkeep.", exc_info=True)
+                    self._handle_update_failure(issue_update, err)
+                    return False
+                except Exception as e:
+                    issue_update.logger.exception(f"Failed to update Redmine issue during upkeep: {e}")
+                    self._handle_update_failure(issue_update, e)
+                    return False
+            else:
+                issue_update.logger.info("No changes detected after all transformations. No Redmine update sent.")
```

Comment:
You could also define the env. var `PYTHONUNBUFFERED` and avoid explicit flushing.